src/Compiler/Service/TransparentCompiler.fs

```diff
@@ -0,0 +1,1647 @@
+                                         prefixPathOpt,
+                                         TcResultsSink.NoSink,
+                                         tcState,
+                                         parsedInput)
+                                        tcState
+
+                                let tcIntermediate =
+                                    {
+                                        finisher = Finisher(fileNode, finisher)
+                                        moduleNamesDict = tcInfo.moduleNamesDict
+                                        tcDiagnosticsRev = []
+                                        tcDependencyFiles = []
+                                        sink = Unchecked.defaultof<_> // TODO: something nicer
+                                    }
+
+                                node.Return(tcIntermediate))
+                        |> NodeCode.Parallel
+
+                    return!
+                        results
+                        |> Array.filter (
+                            function 
+                            | {finisher = Finisher(node=NodeToTypeCheck.ArtificialImplFile idx)} -> 
+                                // When we have the actual PhysicalFile in the same layer as the corresponding ArtificialImplFile then we don't apply the artificial's file finisher because it's not needed and it would lead to doubling of the results 
+                                not (layer.Contains (NodeToTypeCheck.PhysicalFile (idx + 1)))
+                            | _ -> true)
+                        |> mergeIntermediateResults tcInfos
+                        |> processLayer rest
+            }
+
+        node {
+            let! tcInfo = processLayer layers [ bootstrapInfo.InitialTcInfo ]
+            return tcInfo, graph
+        }
+
+    // Type check everything that is needed to check given file
+    let ComputeTcPrior (file: FSharpFile) (bootstrapInfo: BootstrapInfo) (projectSnapshot: FSharpProjectSnapshot) =
+
+        let priorSnapshot = projectSnapshot.UpTo file.Source.FileName
+        let key = priorSnapshot.Key
+
+        TcPriorCache.Get(
+            key,
+            node {
+                use _ =
+                    Activity.start "ComputeTcPrior" [| Activity.Tags.fileName, file.Source.FileName |> Path.GetFileName |]
+
+                // parse required files
+                let files =
+                    seq {
+                        yield! bootstrapInfo.SourceFiles |> Seq.takeWhile ((<>) file)
+                        file
+                    }
+
+                let! parsedInputs = files |> Seq.map (ComputeParseFile bootstrapInfo) |> NodeCode.Parallel
+
+                let! graph, dependencyFiles =
+                    ComputeDependencyGraphForFile priorSnapshot (parsedInputs |> Array.map p13) bootstrapInfo.TcConfig
+
+                // layers that can be processed in parallel
+                let layers = Graph.leafSequence graph |> Seq.toList
```

Comment:
Processing nodes in layers might make scheduling simpler, but it does reduce parallelisation compared with processing each node as soon as all its dependencies have finished.

Was this a conscious decision? It's fine for an MVP, but with potential for improvement 🙂

Reply:
I just wanted to make it as simple as possible and to get it working. But in the end if we can make it faster we should. Might be nice to reuse the scheduling you made here if it's possible. 

But I want to make it work properly first before doing much optimization.

Reply:
Actually, it's not so simple anyway. I was having some issues with duplication from ArtificialImplFiles. So I tried to plug in your graph processing and it seems to work fine. I duplicated some of it for now because of NodeCode/Async, but maybe we'll be able to unify it.